modules/react/collection/lib/useBaseListModel.tsx

```diff
@@ -112,8 +115,17 @@ export const useBaseListModel = createModelHook({
 })(config => {
   const id = useUniqueId(config.id);
 
+  // Optimization to not redo items when `getId` and `getTextValue` references change. They will not
+  // likely change during the lifecycle and we don't want to recalculate items when a lamba is
+  // passed instead of a stable reference.
+  const getIdRef = React.useRef(defaultGetId);
+  const getTextValueRef = React.useRef(defaultGetTextValue);
```

Comment:
nice! you're a master at refs, I feel like this would be a good topic of discussion/ learning. How to use refs more effectively

Reply:
I remember this being an issue for AngularJS as well. The rule was ["always have a 'dot' in your ngModel"](https://github.com/angular/angular.js/wiki/Understanding-Scopes).

It comes from how references and values work in JavasScript. `React.useRef` gives you access to a reference to a value - think of it as a gift that is wrapping a value. The reference to the gift itself never changes, but what's inside could be different every time. I think that's why they choose the name "ref" because it means a reference. Like the literal reference in memory that points to a mutable value.

In JavaScript, there's references and there's values. When you pass data around in JavaScript, the engine makes a decision on how the thing is passed around based on its type. Primitives like strings and numbers are passed by value (a copy) while everything else is by a reference. A reference refers to an identifier/pointer/location that then has to be looked up.

A React ref wraps everything in an object, forcing the user of that thing to do a lookup. This allows the ref to be passed via closures. Closures is how React hooks operate - a closure is a "stack" of scoped variables every time a function or scope is created. A closure access outer variables:

```ts
const foo = 'bar';

const onClick = () => {
  console.log(foo) // `foo` is "closed over" or `onClick` creates a closure that has `foo` in scope
}
```

But what happens here?
```ts
const foo = props.foo;

const onClick = React.useCallback(() => {
  console.log(foo) // what will be logged?
}, [])
```

`React.useCallback`, `React.useMemo`, `React.useEffect`, `React.useLayoutEffect` create memoized closures that are based on a dependency array of reference/value comparisons. In this example, if `props.foo` changes, the `console.log` will log out the old value (the outdated closed over value). This is why all these functions have dependency arrays so they can be re-evaluated. In the case of `use{Layout}Effect`, we want to do something when specific data changes. But this violates rules of hooks. If you create a reference using `React.useRef`, rules of hooks does not complain about a dependency missing from the dependency list because React knows `React.useRef` creates a reference wrapper:

```ts
const foo = React.useRef(props.foo);
foo.current = props.foo

const onClick = React.useCallback(() => {
  console.log(foo.current) // what will be logged?
}, [])
```

Now the reference of `onClick` is stable (will not change each render) AND the console.log will always reflect the latest changes to `props.foo`.